### DISCUSSION TITLE
Rules for emitted events

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/flow#165] [Ideas]
I think there are a few inconsistencies among the events emitted. So I decided to make rules for them.

### Rules
1. All events must contain `streamId` as the first entry and it must always be indexed.
2. If there is a transfer of asset, the counterparty, which is sending/receiving the asset, must be indexed.
4. As an exception, create must have `asset`, `streamId` and `sender` indexed, regardless.
5. No indexing otherwise.

Here is the revised version:

```solidity
event AdjustFlowStream(
    uint256 indexed streamId, uint128 amountOwed, uint128 newRatePerSecond, uint128 oldRatePerSecond
);

event CreateFlowStream(
    uint256 indexed streamId,
    IERC20 indexed asset,
    address indexed sender,
    address recipient,
    uint40 lastTimeUpdate,
    uint128 ratePerSecond
);

event DepositFlowStream(
    uint256 indexed streamId, address indexed funder, uint128 depositAmount
);

event PauseFlowStream(
    uint256 indexed streamId, address recipient, address sender, uint128 amountOwed
);

event RefundFromFlowStream(
    uint256 indexed streamId, address indexed sender, uint128 refundAmount
);

event RestartFlowStream(
    uint256 indexed streamId, address recipient, address sender, uint128 ratePerSecond
);

event VoidFlowStream(
    uint256 indexed streamId, address recipient, address sender, uint128 forfeitedDebt, uint128 newAmountOwed
);

event WithdrawFromFlowStream(
    uint256 indexed streamId, address indexed to, uint128 withdrawnAmount
);
```

RFC @andreivladbrg.

**Comment by andreivladbrg** [comment]
I pretty much agree with your proposed rules and don't have any objections.

Let's also see @razgraf 's opinion since the events are relevant to the subgraph.
Question, would it be useful to have the `asset` indexed in each event?

**Reply by razgraf** [reply]
Regarding the `asset` it's only important for the initialization events (e.g. `CreateFlowStream`). It doesn't change over the course of the stream, so we only need to index it in one event (`create`, the first one). This is generally true for any such system (including lockup, airstreams).

**Reply by razgraf** [reply]
Regarding the rules suggested by @smol-ninja , they seem great! If indexing `asset` provides for some extra consistency, we can keep it (although in data indexers, it's not used for most events).

**Reply by smol-ninja** [reply]
Thanks for your feedback. 

1. Lets not index `asset` then except in create.
2. In create, to be able to index `asset`, we will have to un-index either of  `streamId`, `recipient` or `sender` because only 3 are allowed. Which one is the least useful?

**Reply by smol-ninja** [reply]
> although in data indexers, it's not used for most events

Then if we remove `asset` from the events, will that be OK? My argument is why to include any data if its not important.

**Reply by andreivladbrg** [reply]
since the `ERC721` also indexes the ID minted and `to` (i.e., recipient), between these two, i would un-index the recipient. wdyt?

**Reply by smol-ninja** [reply]
In create, yes.

**Comment by smol-ninja** [comment]
@andreivladbrg updated the OP. Lmk if it looks good, I will create an issue and handle it in my next PR.

**Comment by smol-ninja** [comment]
### Issue
- https://github.com/sablier-labs/flow/issues/169
